src/herder/HerderSCPDriver.cpp

```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
 {
     mSCPMetrics.mAcceptedCommit.Mark();
 }
+
+void
+HerderSCPDriver::setMetricStartTime(uint64_t slotIndex,
+                                    VirtualClock::time_point start,
+                                    bool isNomination)
+{
+    SCPTimingPtr timing;
+    auto it = mSCPExecutionTimes.find(slotIndex);
+    timing = it != mSCPExecutionTimes.end() ? it->second
+                                            : std::make_shared<SCPTiming>();
+
+    if (isNomination)
+    {
+        timing->mNominationStart =
+            make_optional<VirtualClock::time_point>(start);
+    }
+    else
+    {
+        timing->mPrepareStart = make_optional<VirtualClock::time_point>(start);
+    }
+
+    // Update timings map
+    mSCPExecutionTimes[slotIndex] = timing;
+}
+
+void
+HerderSCPDriver::recordSCPExecutionMetrics(uint64_t slotIndex)
+{
+    auto externalizeStart = mApp.getClock().now();
+
+    auto SCPTimingIt = mSCPExecutionTimes.find(slotIndex);
+    if (SCPTimingIt == mSCPExecutionTimes.end())
+    {
+        // No metric was recorded
+        CLOG(WARNING, "Herder")
+            << "No SCP metrics recorded for slot, " << slotIndex
+            << " this should not have happened.";
+        return;
+    }
+
+    auto SCPTiming = SCPTimingIt->second;
+
+    auto checkThreshold = [&](std::chrono::nanoseconds delta) -> bool {
+        if (delta < Herder::NOMINATION_THRESHOLD_NANOSECS)
```

Comment:
the code would be tighter if instead you were computing the threshold at the beginning of the function (outside of the lambda) as single node just means that the threshold is 0